crates/matrix-sdk-crypto/src/identities/user.rs

```diff
@@ -737,7 +763,21 @@ impl OtherUserIdentityData {
         &self.self_signing_key
     }
 
-    /// Pin the current identity
+    /// Remember this identity, ensuring it does not result in a pin violation.
+    ///
+    /// When we first see a user, we assume their cryptographic identity has not
+    /// been tampered with by the homeserver or another entity with
+    /// man-in-the-middle capabilities. We remember this identity and call this
+    /// action "pinning".
+    ///
+    /// If the identity presented for the user changes later on, the newly
+    /// presented identity is considered to be in "pin violation". This
+    /// method explicitly accepts the new identity, allowing it to replace
+    /// the previously pinned one and bringing it out of pin violation.
+    ///
+    /// An identity which is not verified and is in pin violation will require
+    /// user approval before it can be used. See
```

Comment:
Yep, we should, though I'm out of my depth here since I've not been keeping this close of an eye on the new trust requirement code.

Reply:
We don't require the violation to be resolved before using the identity anymore (using here is for sharing room key or trust requirement for decryption). It was considered but it is per design a non blocking warning, so the sdk remembers if the warning was displayed to the user, until resolved the warning should be shown by UI.
So just outdated doc?

Reply:
Hmm, I have a number of questions now.

- So what exactly is the effect of `identity_needs_user_approval() == true`? It simply displays a warning with no other effect?
- Do we differentiate between a `identity_needs_user_approval() == true` state created due to a verified identity mismatch vs a pin violation? Or is the effect exactly the same in both cases?
- Is any of this going to change after the "transition period"?

Also, can you suggest wording that would make it accurate?

Reply:
> What exactly is the effect of identity_needs_user_approval() == true? It simply displays a warning with no other effect?

Yes. (This is for _unverified_ identities.)

> So what is the behaviour when a verified identity is mismatched?

In legacy/transition mode (they are the same) we will block sending and decrypting messages to/from an identity with a verification violation. I think it will be the same in future too.

> what is the name of the method which enforces this case?

`withdraw_verification` changes a (potentially-violating) identity back into a pinned identity. `has_verification_violation` asks whether this identity has a verification violation.

I find these methods very confusing and am trying to think about how to structure them better.

> Is any of this going to change after the "transition period"?

I don't think so.

> Also, can you suggest wording that would make it accurate?

I went for "UIs should display a warning to the user when encountering an identity which is not verified and is in pin violation."